src/NServiceBus.Core/DataBus/FileShareDataBusImplementation.cs

```diff
@@ -53,15 +55,10 @@ public string Put(Stream stream, TimeSpan timeToBeReceived)
 
 			Directory.CreateDirectory(Path.GetDirectoryName(filePath));
 
-			using (var output = new FileStream(filePath, FileMode.CreateNew, FileAccess.Write, FileShare.Read))
+			using (var output = new FileStream(filePath, FileMode.CreateNew, FileAccess.Write, FileShare.Read, 4096, FileOptions.Asynchronous))
 			{
-				var buffer = new byte[32 * 1024];
-				int read;
-
-				while ((read = stream.Read(buffer, 0, buffer.Length)) > 0)
-				{
-					output.Write(buffer, 0, read);
-				}
+			    const int bufferSize = 32 * 1024;
+                await stream.CopyToAsync(output, bufferSize);
```

Comment:
Yes, that's the one I'm referring to. Since we're not writing into the stream ourselves (the source stream does), it becomes source stream responsibility I suspect. Tests didn't fail (parallel reading of file), as we discussed, so that was my indication that it's working as needed.